fix(organizations): allow org owners and admins to edit their own use…

antcybersec opened the pull request:
Allows org owners/admins to see the edit (ellipsis) menu on their own row by permitting self-edit when user is admin/owner. Changes limited to UserListTable.tsx and MemberList.tsx. Fixes #24377.

https://github.com/user-attachments/assets/4eb6fe70-d0ed-4a4e-918c-ff3723c6c3f3



here is the new video ...

https://github.com/user-attachments/assets/094401b4-9b20-4a99-85b8-9b645f503820

antcybersec committed:
fix(organizations): allow org owners and admins to edit their own usernames

- Fix UserListTable to show edit option for org owners/admins on themselves
- Fix MemberList to show edit option for org owners/admins on themselves
- Resolves issue where org owners/admins couldn't see triple dot menu to edit their own usernames
- Maintains existing security by only allowing edit for users with proper permissions

Fixes #24377

dhairyashiil reviewed (changes requested):
Please attach before and after loom video, making it draft until then.

antcybersec committed:
fix(organizations): self-owner sees Edit + Leave; hide Resend/Remove for self; show public link for pending members

antcybersec committed:
fix(organizations): improve UX with confirmation dialogs and proper guards

- Add confirmation dialog for Leave team action
- Prevent last owner from leaving team
- Add proper guards for mobile Remove action
- Fix public page links with username validation
- Disable buttons during pending operations

antcybersec committed:
fix(organizations): address CodeRabbit suggestions

- Use ConfirmationDialogContent for leave dialog consistency
- Add server-side last owner guard in leaveTeamMembership
- Prevent last owner from leaving team at database level

antcybersec committed:
fix(organizations): address CodeRabbit race condition and consistency issues

- Use transaction to prevent race condition in leaveTeamMembership
- Use MembershipRole.OWNER enum instead of string literals
- Use TRPCError for consistent error handling
- Use logger instead of console.log

dhairyashiil reviewed (changes requested):
Could you please check the following?

1. Can you change the role of the owner to admin or member?
2. Also, could you click on “Leave Team”? I mean, is the owner able to leave the team? Please check this as well. ( I think you have added new check for this in your code, right?)

Record both scenarios and please attach a video here.

Additionally, the tests are failing, and there’s a merge conflict. Please fix both.

antcybersec committed:
fix(organizations): allow owner to change own role and leave team

- Remove ownersCount > 1 restriction for leave team button
- Remove server-side last owner restriction for leave team
- Allow owners to change their own role (even if last owner)
- Update both legacy and PBAC role managers for self-role changes
- Remove unused ownersCount variable

antcybersec commented on packages/app-store/salesforce/lib/graphql/SalesforceGraphQLClient.ts:
```diff
@@ -26,14 +25,21 @@ export class SalesforceGraphQLClient {
+          initialDelayMs: Number(process.env.SALESFORCE_GRAPHQL_DELAY_MS),
+          maxDelayMs: Number(process.env.SALESFORCE_GRAPHQL_MAX_DELAY_MS),
+          randomDelay: true,
+          maxNumberAttempts: Number(process.env.SALESFORCE_GRAPHQL_MAX_RETRIES),
```
@dhairyashiil hey please check once i have removed the unnnecesary files ...

antcybersec committed:
fix: update leaveTeamMembership tests to properly mock transaction

- Mock prisma.$transaction to handle transaction-based calls
- Update test setup to mock tx.membership methods instead of direct prismaMock
- Ensure tests properly verify transaction behavior
- Add eslint-disable comments for necessary any types

antcybersec commented:
Hey @dhairyashiil , thanks for reopening the PR.

I see there are new merge conflicts in MemberList.tsx and UserListTable.tsx. Before I start resolving them, just wanted to confirm:

→ Should I rebase my branch on top of the latest main and push the updated version?  
→ Or would you prefer I merge main into the feature branch?

Once I get your confirmation, I’ll resolve the conflicts and re-run the tests so the checks can pass.

Let me know — happy to fix everything needed.

antcybersec committed:
Merge branch 'main' into fix/org-owners-edit-own-username

antcybersec committed:
Update packages/features/pbac/services/legacy-role-manager.service.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antcybersec committed:
Update packages/features/pbac/services/pbac-role-manager.service.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antcybersec committed:
Update packages/lib/server/service/teamService.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

pallava-joshi committed:
Merge branch 'main' into fix/org-owners-edit-own-username

antcybersec committed:
Update packages/lib/server/service/teamService.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antcybersec committed:
Update packages/features/ee/teams/components/MemberList.tsx

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

Ryukemeister reviewed (commented):
hi @antcybersec, thank for you contribution. i'm gonna close this in favour of https://github.com/calcom/cal.com/pull/24407 since we think the fix is implemented in a much better way there, thanks again!
